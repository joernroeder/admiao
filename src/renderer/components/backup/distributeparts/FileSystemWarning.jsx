import React from 'react'

import {
	DistributionTypes,
	useDistributionDispatch,
} from '../../../store/DistributionStore'

import { Flex } from 'reflexbox'
import Button from '../../Button'
import GridWrap from '../../GridWrap'
import Cell from '../../Cell'
import SubHeading from '../../SubHeading'
import Text from '../../Text'
import styled from '@emotion/styled'
import { background, space } from 'styled-system'
import {
	verticalRhythm,
	verticalRhythmProps,
} from '../../../utils/styled-system-rhythm'

const StripedBackground = styled(Flex)`
	height: 100vh;
	${props => {
		const color = props.theme.colors.lightGray
		const size = 8

		console.log(size)

		return `background: repeating-linear-gradient(
		-45deg,
		${color},
		${color} ${props.theme.rhythm(size)},
		transparent ${props.theme.rhythm(size)},
		transparent ${props.theme.rhythm(size * 2)}
	);`
	}};
`

const StripedTextPadding = styled.span`
	//${background};
	${space};
	box-decoration-break: clone;
`

StripedTextPadding.defaultProps = {
	background: 'white',
	px: 2,
	pt: 0,
	pb: 1,
}

const StripedBoxPadding = styled.span`
	//${background};
	${verticalRhythm};
	${space};
	display: inline-block;
`

StripedBoxPadding.defaultProps = {
	background: 'white',
	pt: 0,
	px: 1,
	pb: 2,
	...verticalRhythmProps(0),
}

const FileSystemWarning = () => {
	const dispatch = useDistributionDispatch()

	return (
		<StripedBackground>
			<GridWrap flexDirection={'column'} backgroundColor={'transparent'}>
				<Cell gridColumn={6} gridOffset={4}>
					<SubHeading mt={8}>
						<StripedTextPadding>
							Be careful â€“ you are about to leave the safe path!
						</StripedTextPadding>
					</SubHeading>
				</Cell>
				<Cell gridColumn={4} gridOffset={5}>
					<Text mt={5} backgroundColor={'white'} px={1} py={1}>
						By handling generated shares yourself you may expose
						yourself to risk accidentally.
						<br />
						Are you sure you want to continue?
					</Text>
				</Cell>
				<Cell gridColumn={4} gridOffset={5} mt={3}>
					<StripedBoxPadding>
						<Button
							to={'../'}
							mr={2}
							variant={'filled'}
							showArrow={false}
						>
							No, go back
						</Button>
						<Button
							to={'../../enter-seed'}
							onClick={() =>
								dispatch({
									type: DistributionTypes.FILE_SYSTEM,
								})
							}
						>
							Yes, i take the risk
						</Button>
					</StripedBoxPadding>
				</Cell>
			</GridWrap>
		</StripedBackground>
	)
}

export default FileSystemWarning
