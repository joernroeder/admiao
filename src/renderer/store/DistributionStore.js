import React, { createContext, useContext, useReducer, useEffect } from 'react'

const localStorageKey = 'DistributionState'

const initialState = {
	distributionType: undefined,
	distributionIdentifier: undefined,
}

const DistributionTypes = {
	LOCAL_PRINT: 'LOCAL_PRINT',
	REMOTE_PRINT: 'REMOTE_PRINT',
	DIGITAL: 'DIGITAL',
	FILE_SYSTEM: 'FILE_SYSTEM',
}

function reducer(state, action) {
	// sets Distribution Type
	if (Object.keys(DistributionTypes).includes(action.type)) {
		return {
			...state,
			distributionType: DistributionTypes[action.type],
			distributionIdentifier: action.payload || undefined,
		}
	}

	switch (action.type) {
		case 'RESET': {
			localStorage.removeItem(localStorageKey)

			return {
				...initialState,
			}
		}

		case 'SET_IDENTIFIER': {
			return {
				...state,
				distributionIdentifier: action.payload,
			}
		}

		default:
			return state
	}
}

const DistributionStateContext = createContext(undefined)
const DistributionDispatchContext = createContext(undefined)

function DistributionProvider({ children }) {
	const [state, dispatch] = useReducer(
		reducer,
		initialState,
		initialState => {
			const previousState = localStorage.getItem(localStorageKey)

			return previousState ? JSON.parse(previousState) : initialState
		}
	)

	// store distributionType in local storage.
	// This stores the distribution type to improve the backup process.
	useEffect(() => {
		if (state.isInitialState) {
			return
		}

		// filter out empty and initial states
		if (!Object.values(state).filter(Boolean).length) {
			return
		}

		console.log(Object.values(state).filter(Boolean).length)
		localStorage.setItem(localStorageKey, JSON.stringify(state))
	}, [state])

	return (
		<DistributionStateContext.Provider value={state}>
			<DistributionDispatchContext.Provider value={dispatch}>
				{children}
			</DistributionDispatchContext.Provider>
		</DistributionStateContext.Provider>
	)
}

function useDistributionState() {
	const context = useContext(DistributionStateContext)

	if (context === undefined) {
		throw new Error(
			'useDistributionState must be used within a DistributionProvider'
		)
	}
	return context
}

function useDistributionDispatch() {
	const context = useContext(DistributionDispatchContext)

	if (context === undefined) {
		throw new Error(
			'useDistributionDispatch must be used within a DistributionProvider'
		)
	}
	return context
}

export {
	DistributionProvider,
	useDistributionState,
	useDistributionDispatch,
	DistributionTypes,
}
