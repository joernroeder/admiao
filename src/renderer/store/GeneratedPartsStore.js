import React, { createContext, useContext, useReducer } from 'react'

const initialState = {
	parts: [],
	confirmed: [],
}

function reducer(state, action) {
	switch (action.type) {
		case 'SET_PARTS': {
			console.log('setting parts in store', action)
			return {
				...state,
				confirmed: Array(action.payload.length).fill(false),
				parts: action.payload,
			}
		}

		case 'CONFIRMED': {
			const updatedConfirmed = {
				confirmed: state.confirmed.map((isConfirmed, index) => {
					if (index === action.payload) {
						return true
					}

					return isConfirmed
				}),
			}

			return {
				...state,
				...updatedConfirmed,
			}
		}

		case 'RESET': {
			return {
				...initialState,
			}
		}

		default:
			return state
	}
}

const GeneratedPartsStateContext = createContext(undefined)
const GeneratedPartsDispatchContext = createContext(undefined)

function GeneratedPartsProvider({ children }) {
	const [state, dispatch] = useReducer(reducer, initialState)

	return (
		<GeneratedPartsStateContext.Provider value={state}>
			<GeneratedPartsDispatchContext.Provider value={dispatch}>
				{children}
			</GeneratedPartsDispatchContext.Provider>
		</GeneratedPartsStateContext.Provider>
	)
}

function useGeneratedParts() {
	const context = useContext(GeneratedPartsStateContext)

	if (context === undefined) {
		throw new Error(
			'useGeneratedParts must be used within a GeneratedPartsProvider'
		)
	}
	return context
}

function useGeneratedPartsDispatch() {
	const context = useContext(GeneratedPartsDispatchContext)

	if (context === undefined) {
		throw new Error(
			'useDistributionDispatch must be used within a GeneratedPartsProvider'
		)
	}
	return context
}

export { GeneratedPartsProvider, useGeneratedParts, useGeneratedPartsDispatch }
