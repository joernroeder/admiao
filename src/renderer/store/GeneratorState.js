import React, { createContext, useContext, useState } from 'react'

const STATES = {
	loading: 'LOADING',
	successfullyGenerated: 'SUCCESSFULLY_GENERATED',
	generationFailed: 'GENERATION_FAILED',
	successfullyDistributed: 'SUCCESSFULLY_DISTRIBUTED',
	distributionFailed: 'DISTRIBUTION_FAILED',
}

const GeneratorStateContext = createContext(undefined)

function GeneratorStateProvider({ children }) {
	const [generatorState, setGeneratorState] = useState({
		state: undefined,
		pendingIdentifier: null,
		error: null,
		// additional data returned by the ipc methods may get stored here.
		// todo be more precise/refactor once all distribution types are implemented
		meta: {},
	})

	return (
		<GeneratorStateContext.Provider
			value={[generatorState, setGeneratorState]}
		>
			{children}
		</GeneratorStateContext.Provider>
	)
}

function useGeneratorState() {
	const context = useContext(GeneratorStateContext)

	if (context === undefined) {
		throw new Error(
			'useGeneratedParts must be used within a GeneratedPartsProvider'
		)
	}
	return context
}

export { GeneratorStateProvider, useGeneratorState, STATES }
