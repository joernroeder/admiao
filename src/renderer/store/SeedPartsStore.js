import React, { createContext, useContext, useEffect, useReducer } from 'react'

const localStorageKey = 'SeedPartsState'

const initialState = {
	uniquePartsN: 3,
	requiredPartsT: 2,

	partsPerDisk: 1,
	requiredDisks: 3,

	recoveredFromPreviousSelection: false,
	isInitialState: true,
}

const calcDisks = (uniquePartsN, requiredPartsT) => {
	const partsPerDisk = requiredPartsT - 1
	const requiredDisks = Math.ceil(uniquePartsN / partsPerDisk)

	return {
		partsPerDisk,
		requiredDisks,
	}
}

function reducer(state, action) {
	const { uniquePartsN, requiredPartsT } = state

	const isModifiedState = {
		recoveredFromPreviousSelection: false,
		isInitialState: false,
	}

	switch (action.type) {
		case 'INCREMENT_REQUIRED': {
			const updatedRequiredPartsT = Math.min(
				requiredPartsT + 1,
				uniquePartsN
			)

			return {
				...state,
				requiredPartsT: updatedRequiredPartsT,
				...calcDisks(state.uniquePartsN, updatedRequiredPartsT),
				...isModifiedState,
			}
		}

		case 'DECREMENT_REQUIRED': {
			const updatedRequiredPartsT = Math.max(requiredPartsT - 1, 1)

			return {
				...state,
				requiredPartsT: updatedRequiredPartsT,
				...calcDisks(state.uniquePartsN, updatedRequiredPartsT),
				...isModifiedState,
			}
		}

		case 'INCREMENT_UNIQUE': {
			const updatedUniquePartsN = uniquePartsN + 1

			return {
				...state,
				uniquePartsN: updatedUniquePartsN,
				...calcDisks(updatedUniquePartsN, state.requiredPartsT),
				...isModifiedState,
			}
		}

		case 'DECREMENT_UNIQUE': {
			const updatedUniquePartsN = Math.max(uniquePartsN - 1, 1)
			// also decrement required parts if they would exceed otherwise
			const updatedRequiredPartsT = Math.min(
				updatedUniquePartsN,
				requiredPartsT
			)

			return {
				...state,
				uniquePartsN: updatedUniquePartsN,
				requiredPartsT: updatedRequiredPartsT,
				...calcDisks(updatedUniquePartsN, updatedRequiredPartsT),
				...isModifiedState,
			}
		}

		case 'CONFIRM': {
			return {
				...state,
				...isModifiedState,
			}
		}

		case 'RESET': {
			localStorage.removeItem(localStorageKey)

			return {
				...initialState,
			}
		}

		default:
			return state
	}
}

const SeedPartsStateContext = createContext(undefined)
const SeedPartsDispatchContext = createContext(undefined)

function SeedPartsProvider({ children }) {
	// uses a previous state from local storage if available or uses the initial one defined above.
	const [state, dispatch] = useReducer(
		reducer,
		initialState,
		initialState => {
			const previousState = localStorage.getItem(localStorageKey)

			return previousState
				? {
						...JSON.parse(previousState),
						recoveredFromPreviousSelection: true,
				  }
				: initialState
		}
	)

	// store current state in local storage.
	// This is used to improve the backup process if the user paused to print the
	// seed templates.
	useEffect(() => {
		if (state.isInitialState) {
			return
		}

		localStorage.setItem(localStorageKey, JSON.stringify(state))
	}, [state])

	return (
		<SeedPartsStateContext.Provider value={state}>
			<SeedPartsDispatchContext.Provider value={dispatch}>
				{children}
			</SeedPartsDispatchContext.Provider>
		</SeedPartsStateContext.Provider>
	)
}

function useSeedPartsState() {
	const context = useContext(SeedPartsStateContext)

	if (context === undefined) {
		throw new Error(
			'useSeedPartsState must be used within a SeedPartsProvider'
		)
	}
	return context
}

function useSeedPartsDispatch() {
	const context = useContext(SeedPartsDispatchContext)

	if (context === undefined) {
		throw new Error(
			'useSeedPartsDispatch must be used within a SeedPartsProvider'
		)
	}
	return context
}

export { SeedPartsProvider, useSeedPartsState, useSeedPartsDispatch }
