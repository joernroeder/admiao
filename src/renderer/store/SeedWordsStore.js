import React, { createContext, useContext, useReducer } from 'react'
const bip39 = require('bip39')

const initialWords = [
	/*
	'answer',
	'dice',
	'warm',
	'cup',
	'shock',
	'orbit',
	'pistol',
	'seat',
	'modify',
	'matrix',
	'genre',
	'renew',
	*/
]

const initialState = {
	words: initialWords,
	total: 12, // todo handle other seeds 18, 24
	isConfirmed: false,
	isValid: getIsValid(initialWords),
}

function getIsValid(words) {
	return bip39.validateMnemonic(words.join(' '))
}

function wordsReducer(state, action) {
	const { words } = state

	switch (action.type) {
		case 'APPEND': {
			const updatedWords = [...words, action.payload]

			return {
				...state,
				words: updatedWords,
				isValid: getIsValid(updatedWords),
			}
		}

		case 'RESET': {
			return {
				...state,
				words: [],
				isConfirmed: false,
				isValid: false,
			}
		}

		case 'CLEANUP_SEED': {
			return {
				...state,
				words: [],
			}
		}

		case 'POP': {
			const updatedWords = [...words.slice(0, -1)]

			return {
				...state,
				words: updatedWords,
				isValid: getIsValid(updatedWords),
			}
		}

		case 'CONFIRM': {
			return {
				...state,
				isConfirmed: true,
			}
		}

		default:
			return state
	}
}

const SeedWordsStateContext = createContext(undefined)
const SeedWordsDispatchContext = createContext(undefined)

function SeedWordsProvider({ children }) {
	const [state, dispatch] = useReducer(wordsReducer, initialState)

	return (
		<SeedWordsStateContext.Provider value={state}>
			<SeedWordsDispatchContext.Provider value={dispatch}>
				{children}
			</SeedWordsDispatchContext.Provider>
		</SeedWordsStateContext.Provider>
	)
}

function useSeedWordsState() {
	const context = useContext(SeedWordsStateContext)

	if (context === undefined) {
		throw new Error(
			'useSeedWordsState must be used within a SeedWordsProvider'
		)
	}
	return context
}

function useSeedWordsDispatch() {
	const context = useContext(SeedWordsDispatchContext)

	if (context === undefined) {
		throw new Error(
			'useSeedWordsDispatch must be used within a SeedWordsProvider'
		)
	}
	return context
}

export {
	SeedWordsProvider,
	useSeedWordsState,
	useSeedWordsDispatch,
	wordsReducer,
}
