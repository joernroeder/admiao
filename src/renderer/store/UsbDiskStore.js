import React, { createContext, useContext, useReducer } from 'react'

const initialState = {
	useManagedDisks: true,
}

function reducer(state, action) {
	switch (action.type) {
		case 'USE_PREPARED_DISKS': {
			return {
				...state,
				useManagedDisks: true,
			}
		}

		case 'DISABLE_DISK_MANAGEMENT': {
			return {
				...state,
				useManagedDisks: false,
			}
		}

		default:
			return state
	}
}

const UsbDiskStateContext = createContext(undefined)
const UsbDiskDispatchContext = createContext(undefined)

function UsbDiskProvider({ children }) {
	const [state, dispatch] = useReducer(reducer, initialState)

	return (
		<UsbDiskStateContext.Provider value={state}>
			<UsbDiskDispatchContext.Provider value={dispatch}>
				{children}
			</UsbDiskDispatchContext.Provider>
		</UsbDiskStateContext.Provider>
	)
}

function useUsbDiskState() {
	const context = useContext(UsbDiskStateContext)

	if (context === undefined) {
		throw new Error('useUsbDiskState must be used within a UsbDiskProvider')
	}
	return context
}

function useUsbDiskDispatch() {
	const context = useContext(UsbDiskDispatchContext)

	if (context === undefined) {
		throw new Error(
			'useUsbDiskDispatch must be used within a UsbDiskProvider'
		)
	}
	return context
}

export { UsbDiskProvider, useUsbDiskState, useUsbDiskDispatch }
