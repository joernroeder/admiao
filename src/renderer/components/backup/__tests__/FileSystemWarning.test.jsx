import React from 'react'

import { fireEvent, render } from '@testing-library/react'
import { ThemeProvider } from 'emotion-theming'
import { DistributionProvider } from '../../../store/DistributionStore'

import FileSystemWarning from '../distributeparts/FileSystemWarning'

describe('File-System Warning', () => {
	const renderWithProviders = (children, renderFunction = render) => {
		const theme = { colors: { lightGray: '#ccc' }, rhythm: size => size }

		return renderFunction(
			<ThemeProvider theme={theme}>
				<DistributionProvider>{children}</DistributionProvider>
			</ThemeProvider>
		)
	}

	test('it should render correctly', () => {
		const { getByText, container } = renderWithProviders(
			<FileSystemWarning />
		)

		expect(getByText(/be careful/i, { exact: false })).toBeInTheDocument()

		expect(container.firstChild).toMatchSnapshot()
	})

	test('it should correctly navigate to `/` on back button click', () => {
		const { getByText, history } = renderWithProviders(
			<FileSystemWarning />,
			global.renderWithRouter
		)

		fireEvent.click(getByText('No, go back'))

		expect(history.location.pathname).toEqual('/')
	})

	test('it should correctly navigate to `/enter-seed` on forward button click', () => {
		const { getByText, history } = renderWithProviders(
			<FileSystemWarning />,
			global.renderWithRouter
		)

		fireEvent.click(getByText('Yes, i take the risk'))

		expect(history.location.pathname).toEqual('/enter-seed')
	})
})
