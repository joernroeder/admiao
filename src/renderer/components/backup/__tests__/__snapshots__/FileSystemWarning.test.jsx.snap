// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`File-System Warning it should render correctly 1`] = `
.emotion-12 {
  box-sizing: border-box;
  margin: 0;
  min-width: 0;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  height: 100vh;
  background: repeating-linear-gradient( -45deg, #ccc, #ccc 8, transparent 8, transparent 16 );
}

.emotion-11 {
  box-sizing: border-box;
  margin: 0;
  min-width: 0;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  background-color: transparent;
  margin-left: -4px;
  margin-right: -4px;
  color: black;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

.emotion-2 {
  box-sizing: border-box;
  margin: 0;
  min-width: 0;
  padding-left: 4px;
  padding-right: 4px;
  width: 50%;
}

.emotion-1 {
  color: black;
  margin-top: 512px;
  margin-bottom: 0;
  font-size: 14px;
  line-height: 1;
  -webkit-transform: 1px;
  -ms-transform: 1px;
  transform: 1px;
  font-family: 'SpaceGrotesk',sans-serif;
  -webkit-letter-spacing: 0.01em;
  -moz-letter-spacing: 0.01em;
  -ms-letter-spacing: 0.01em;
  letter-spacing: 0.01em;
  text-rendering: geometricPrecision;
  -webkit-font-smoothing: antialiased;
  font-variant-ligatures: common-ligatures;
}

@media screen and (min-width:40em) {
  .emotion-1 {
    font-size: 16px;
    line-height: 2;
    -webkit-transform: 2px;
    -ms-transform: 2px;
    transform: 2px;
  }
}

@media screen and (min-width:52em) {
  .emotion-1 {
    font-size: 16px;
    line-height: 2;
    -webkit-transform: 2px;
    -ms-transform: 2px;
    transform: 2px;
  }
}

.emotion-0 {
  padding-left: 8px;
  padding-right: 8px;
  padding-top: 0;
  padding-bottom: 4px;
  -webkit-box-decoration-break: clone;
  box-decoration-break: clone;
}

.emotion-4 {
  box-sizing: border-box;
  margin: 0;
  min-width: 0;
  padding-left: 4px;
  padding-right: 4px;
  width: 33.33333333333333%;
}

.emotion-3 {
  background-color: white;
  color: black;
  margin-top: 64px;
  padding-left: 4px;
  padding-right: 4px;
  padding-top: 4px;
  padding-bottom: 4px;
  margin-bottom: 0;
  font-size: 12px;
  line-height: 0;
  font-size: 12px;
  line-height: 0;
  -webkit-transform: 0;
  -ms-transform: 0;
  transform: 0;
  text-rendering: optimizeLegibility;
  -webkit-font-smoothing: antialiased;
  font-variant-ligatures: common-ligatures;
}

.emotion-10 {
  box-sizing: border-box;
  margin: 0;
  min-width: 0;
  margin-top: 16px;
  padding-left: 4px;
  padding-right: 4px;
  width: 33.33333333333333%;
}

.emotion-9 {
  font-size: 12px;
  line-height: 0;
  -webkit-transform: 0;
  -ms-transform: 0;
  transform: 0;
  padding-top: 0;
  padding-left: 4px;
  padding-right: 4px;
  padding-bottom: 8px;
  display: inline-block;
}

.emotion-6 {
  border-width: 2px;
  border-color: black;
  border-style: solid;
  margin-right: 8px;
  padding-left: 8px;
  padding-right: 8px;
  padding-top: 4px;
  padding-bottom: 4px;
  font-size: 12px;
  line-height: 0;
  -webkit-letter-spacing: 0.025em;
  -moz-letter-spacing: 0.025em;
  -ms-letter-spacing: 0.025em;
  letter-spacing: 0.025em;
  font-size: 12px;
  line-height: 0;
  -webkit-transform: 0;
  -ms-transform: 0;
  transform: 0;
  color: white;
  background-color: black;
  display: inline-block;
  -webkit-text-decoration: none;
  text-decoration: none;
  font-weight: 700;
  box-sizing: border-box;
  position: relative;
  z-index: 10;
  overflow: hidden;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  text-rendering: geometricPrecision;
  -webkit-font-smoothing: antialiased;
  cursor: pointer;
}

.emotion-6 .icon {
  display: inline-block;
  -webkit-transition: -webkit-transform 0.25s ease-in;
  -webkit-transition: transform 0.25s ease-in;
  transition: transform 0.25s ease-in;
}

.emotion-6:hover .icon {
  -webkit-transition-timing-function: ease-out;
  transition-timing-function: ease-out;
  -webkit-transition-duration: 0.15s;
  transition-duration: 0.15s;
}

.emotion-6:hover .icon.next {
  -webkit-transform: translateX(0.4em);
  -ms-transform: translateX(0.4em);
  transform: translateX(0.4em);
}

.emotion-6:hover .icon.prev {
  -webkit-transform: translateX(-0.4em);
  -ms-transform: translateX(-0.4em);
  transform: translateX(-0.4em);
}

.emotion-6:hover .icon.flip {
  -webkit-transform: rotate(90deg);
  -ms-transform: rotate(90deg);
  transform: rotate(90deg);
}

.emotion-6 > span:first-of-type {
  position: relative;
  z-index: 10;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
}

.emotion-5 {
  background-color: whiteTransparent;
  display: block;
  position: absolute;
  top: 0;
  right: 0;
  bottom: 0;
  left: 0;
  opacity: 0;
  -webkit-transform: skew( -14deg, 0 ) translateX( 110% ) translateY( 0% );
  -ms-transform: skew( -14deg, 0 ) translateX( 110% ) translateY( 0% );
  transform: skew( -14deg, 0 ) translateX( 110% ) translateY( 0% );
  -webkit-transition: -webkit-transform 0.25s ease-in,opacity 0.25s ease-in;
  -webkit-transition: transform 0.25s ease-in,opacity 0.25s ease-in;
  transition: transform 0.25s ease-in,opacity 0.25s ease-in;
}

*:hover > .emotion-5,
*:focus > .emotion-5 {
  opacity: 1;
  -webkit-transition-timing-function: ease-out;
  transition-timing-function: ease-out;
  -webkit-transition-duration: 0.15s;
  transition-duration: 0.15s;
  -webkit-transform: translateX( 1.9em ) translateY(0) skew( -14deg, 0 );
  -ms-transform: translateX( 1.9em ) translateY(0) skew( -14deg, 0 );
  transform: translateX( 1.9em ) translateY(0) skew( -14deg, 0 );
}

.emotion-8 {
  border-width: 2px;
  border-color: black;
  border-style: solid;
  padding-left: 8px;
  padding-right: 8px;
  padding-top: 4px;
  padding-bottom: 4px;
  font-size: 12px;
  line-height: 0;
  -webkit-letter-spacing: 0.025em;
  -moz-letter-spacing: 0.025em;
  -ms-letter-spacing: 0.025em;
  letter-spacing: 0.025em;
  font-size: 12px;
  line-height: 0;
  -webkit-transform: 0;
  -ms-transform: 0;
  transform: 0;
  color: black;
  background-color: transparent;
  display: inline-block;
  -webkit-text-decoration: none;
  text-decoration: none;
  font-weight: 700;
  box-sizing: border-box;
  position: relative;
  z-index: 10;
  overflow: hidden;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  text-rendering: geometricPrecision;
  -webkit-font-smoothing: antialiased;
  cursor: pointer;
}

.emotion-8 .icon {
  display: inline-block;
  -webkit-transition: -webkit-transform 0.25s ease-in;
  -webkit-transition: transform 0.25s ease-in;
  transition: transform 0.25s ease-in;
}

.emotion-8:hover .icon {
  -webkit-transition-timing-function: ease-out;
  transition-timing-function: ease-out;
  -webkit-transition-duration: 0.15s;
  transition-duration: 0.15s;
}

.emotion-8:hover .icon.next {
  -webkit-transform: translateX(0.4em);
  -ms-transform: translateX(0.4em);
  transform: translateX(0.4em);
}

.emotion-8:hover .icon.prev {
  -webkit-transform: translateX(-0.4em);
  -ms-transform: translateX(-0.4em);
  transform: translateX(-0.4em);
}

.emotion-8:hover .icon.flip {
  -webkit-transform: rotate(90deg);
  -ms-transform: rotate(90deg);
  transform: rotate(90deg);
}

.emotion-8 > span:first-of-type {
  position: relative;
  z-index: 10;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
}

.emotion-7 {
  background-color: blackTransparent;
  display: block;
  position: absolute;
  top: 0;
  right: 0;
  bottom: 0;
  left: 0;
  opacity: 0;
  -webkit-transform: skew( -14deg, 0 ) translateX( -110% ) translateY( 0% );
  -ms-transform: skew( -14deg, 0 ) translateX( -110% ) translateY( 0% );
  transform: skew( -14deg, 0 ) translateX( -110% ) translateY( 0% );
  -webkit-transition: -webkit-transform 0.25s ease-in,opacity 0.25s ease-in;
  -webkit-transition: transform 0.25s ease-in,opacity 0.25s ease-in;
  transition: transform 0.25s ease-in,opacity 0.25s ease-in;
}

*:hover > .emotion-7,
*:focus > .emotion-7 {
  opacity: 1;
  -webkit-transition-timing-function: ease-out;
  transition-timing-function: ease-out;
  -webkit-transition-duration: 0.15s;
  transition-duration: 0.15s;
  -webkit-transform: translateX( -1.9em ) translateY(0) skew( -14deg, 0 );
  -ms-transform: translateX( -1.9em ) translateY(0) skew( -14deg, 0 );
  transform: translateX( -1.9em ) translateY(0) skew( -14deg, 0 );
}

<div
  class="emotion-12"
>
  <div
    class="emotion-11"
  >
    <div
      class="emotion-2"
    >
      <h2
        class="emotion-1"
        color="black"
        font-size="1,2,2"
      >
        <span
          class="emotion-0"
        >
          Be careful – you are about to leave the safe path!
        </span>
      </h2>
    </div>
    <div
      class="emotion-4"
    >
      <p
        class="emotion-3"
        color="black"
        font-size="0"
      >
        By handling generated shares yourself you may expose yourself to risk accidentally.
        <br />
        Are you sure you want to continue?
      </p>
    </div>
    <div
      class="emotion-10"
    >
      <span
        class="emotion-9"
        font-size="0"
      >
        <a
          aria-current="page"
          class="emotion-6"
          href="/"
        >
          <span>
            <span
              class="icon prev"
            >
              ←
            </span>
             
            No, go back
          </span>
          <span
            class="emotion-5"
          />
        </a>
        <a
          class="emotion-8"
          href="/enter-seed"
        >
          <span>
            Yes, i take the risk
             
            <span
              class="icon next"
            >
              →
            </span>
          </span>
          <span
            class="emotion-7"
          />
        </a>
      </span>
    </div>
  </div>
</div>
`;
